spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarAutoConfiguration.java

```diff
@@ -0,0 +1,178 @@
+import org.springframework.pulsar.core.DefaultTopicResolver;
+import org.springframework.pulsar.core.PulsarAdministration;
+import org.springframework.pulsar.core.PulsarConsumerFactory;
+import org.springframework.pulsar.core.PulsarProducerFactory;
+import org.springframework.pulsar.core.PulsarReaderFactory;
+import org.springframework.pulsar.core.PulsarTemplate;
+import org.springframework.pulsar.core.SchemaResolver;
+import org.springframework.pulsar.core.SchemaResolver.SchemaResolverCustomizer;
+import org.springframework.pulsar.core.TopicResolver;
+import org.springframework.pulsar.function.PulsarFunction;
+import org.springframework.pulsar.function.PulsarFunctionAdministration;
+import org.springframework.pulsar.function.PulsarSink;
+import org.springframework.pulsar.function.PulsarSource;
+import org.springframework.pulsar.observation.PulsarTemplateObservationConvention;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Apache Pulsar.
+ *
+ * @author Soby Chacko
+ * @author Chris Bono
+ * @author Alexander Preuß
+ * @since 3.2
+ */
+@AutoConfiguration
+@ConditionalOnClass(PulsarTemplate.class)
+@EnableConfigurationProperties(PulsarProperties.class)
+@Import({ PulsarAnnotationDrivenConfiguration.class })
+public class PulsarAutoConfiguration {
+
+	private final PulsarProperties properties;
+
+	public PulsarAutoConfiguration(PulsarProperties properties) {
+		this.properties = properties;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public PulsarClientFactoryBean pulsarClientFactoryBean() {
+		return new PulsarClientFactoryBean(this.properties.buildClientProperties());
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConditionalOnProperty(name = "spring.pulsar.producer.cache.enabled", havingValue = "false")
+	public PulsarProducerFactory<?> pulsarProducerFactory(PulsarClient pulsarClient, TopicResolver topicResolver) {
+		return new DefaultPulsarProducerFactory<>(pulsarClient, this.properties.buildProducerProperties(),
+				topicResolver);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConditionalOnClass(Caffeine.class)
+	@ConditionalOnProperty(name = "spring.pulsar.producer.cache.enabled", havingValue = "true", matchIfMissing = true)
+	public PulsarProducerFactory<?> cachingPulsarProducerFactory(PulsarClient pulsarClient,
+			TopicResolver topicResolver) {
+		return new CachingPulsarProducerFactory<>(pulsarClient, this.properties.buildProducerProperties(),
+				topicResolver, this.properties.getProducer().getCache().getExpireAfterAccess(),
+				this.properties.getProducer().getCache().getMaximumSize(),
+				this.properties.getProducer().getCache().getInitialCapacity());
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public PulsarTemplate<?> pulsarTemplate(PulsarProducerFactory<?> pulsarProducerFactory,
+			ObjectProvider<ProducerInterceptor> interceptorsProvider, SchemaResolver schemaResolver,
+			TopicResolver topicResolver,
+			ObjectProvider<io.micrometer.observation.ObservationRegistry> observationRegistryProvider,
```

Comment:
You may notice that we are "sneaking in" the reference to `ObservationRegistry`. I do understand why it is a prohibited import as it belongs under actuator auto-configure. However, the Pulsar template/listener use the registry directly (if provided) to record observations. 

Spring Kafka also uses the registry directly, but rather than constructor injection, it does a bean lookup in [afterSingletonsInstantiated()](https://github.com/spring-projects/spring-kafka/blob/4eacaa9dce252357c7ef4a7632430dc362a3fd8f/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java#L446).  We will change Spring Pulsar to do the same thing shortly. Long term I want to explore instead using a filter or some other mechanism so that it does not have a direct dependency on the registry.

Reply:
This has been fixed (no more observation registry in the auto-configuration.